File: app/src/main/java/com/kode/utair/data/repository/CityRepository.java
package com.kode.utair.data.repository;

import com.kode.utair.data.repository.datasource.CloudCityDataSource;
import com.kode.utair.domain.repository.ICityRepository;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;

public class CityRepository implements ICityRepository {

    private CloudCityDataSource cloudCityDataSource;

    @Inject
    public CityRepository(CloudCityDataSource cloudCityDataSource) {
        this.cloudCityDataSource = cloudCityDataSource;
    }

    @Override
    public Single<List<String>> getCitiesList() {
        return cloudCityDataSource.getCitiesList();
    }

}

File: app/src/main/java/com/kode/utair/data/repository/CityRepository.kt
package com.kode.utair.data.repository

import com.kode.utair.data.network.UTairApiService
import com.kode.utair.data.network.mappers.CitiesListResponseMapper
import com.kode.utair.domain.repository.ICityRepository
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class CityRepository @Inject constructor(
        private val utairApiService: UTairApiService,
        private val citiesListResponseMapper: CitiesListResponseMapper
) : ICityRepository {

    override fun getCitiesList(): Single<List<String>> {
        return utairApiService.getCities()
                .map { citiesListResponseMapper.map(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

}